File: src/main/java/com/wby/thread/manythread/Chapetor15泛型/node8擦除的补偿/child1创建类型实例/InstantiateGenericType.cpp
//: generics/InstantiateGenericType.cpp
// C++, not Java!

template<class T> class Foo {
  T x; // Create a field of type T
  T* y; // Pointer to T
public:
  // Initialize the pointer:
  Foo() { y = new T(); }
};

class Bar {};

int main() {
  Foo<Bar> fb;
  Foo<int> fi; // ... and it works with primitives
} ///:~
/*
在Erased.java中创建一个new T()的尝试无法实现，部分原因是擦除，另一部分就是编译器不能验证T具有默认(无参)构造器。
但在C++中，这种操作很自然、只管，并且很安全(他是在编译期受到检查的)

java中的解决方案是传递一个工厂对象，并使用他来创建新的实例。最方便的工厂对象就是Class对象，因此如果使用类型标签，
那么你就可以使用newInstance方法来创建这个类型的新对象
见InstantiateGenericType.java
*/

File: src/main/java/com/wby/thread/manythread/Chapetor15泛型/node8擦除的补偿/child1创建类型实例/InstantiateGenericType.java
package com.wby.thread.manythread.Chapetor15泛型.node8擦除的补偿.child1创建类型实例;//: generics/InstantiateGenericType.java

import static com.wby.thread.manythread.net.mindview.util.Print.print;

class ClassAsFactory<T> {
  T x;
  public ClassAsFactory(Class<T> kind) {
    try {
      x = kind.newInstance();
    } catch(Exception e) {
      throw new RuntimeException(e);
    }
  }
}

class Employee {}

public class InstantiateGenericType {
  public static void main(String[] args) {
    ClassAsFactory<Employee> fe =
      new ClassAsFactory<Employee>(Employee.class);
    print("ClassAsFactory<Employee> succeeded");
    try {
      ClassAsFactory<Integer> fi =
        new ClassAsFactory<Integer>(Integer.class);
    } catch(Exception e) {
      print("ClassAsFactory<Integer> failed");
    }
  }
} /* Output:
ClassAsFactory<Employee> succeeded
ClassAsFactory<Integer> failed
*///:~
/**
* @Description:  这可以编译，但是会因为ClassAsFactory<Integer>而失败，因为Integer没有任何默认的构造器。因为这个错误不是在编译期捕获的，所以
 * SUn的伙伴们对这种方式并不赞成，他们建议使用显式的工厂，并将限制类型，使得只能接受实现了这个工厂的类：
*/
interface FactoryI<T> {
  T create();
}

class Foo2<T> {
  private T x;
  public <F extends FactoryI<T>> Foo2(F factory) {
    x = factory.create();
  }
  // ...
}

class IntegerFactory implements FactoryI<Integer> {
  public Integer create() {
    return new Integer(0);
  }
}

class Widget {
  public static class Factory implements FactoryI<Widget> {
    public Widget create() {
      return new Widget();
    }
  }
}

class FactoryConstraint {
  public static void main(String[] args) {
    new Foo2<Integer>(new IntegerFactory());
    new Foo2<Widget>(new Widget.Factory());
  }
} ///:~
/**
* @Description: 注意，这确实只是传递Class<T>的一种变体。两种方式都传递了工厂对象，Class<T>碰巧是内建的工厂对象，而上面的方式创建了一个显式的工厂对象，
 * 但是你却获得了编译期检查。
 *
 * 另种方式是 模板方法 设计模式。下面的例子中，get方法是模板方法，而create方法是在子类中定义的、用来产生子类类型的对象：
*/
abstract class GenericWithCreate<T> {
  final T element;
  GenericWithCreate() { element = create(); }
  abstract T create();
}

class X {}

class Creator extends GenericWithCreate<X> {
  X create() { return new X(); }
  void f() {
    System.out.println(element.getClass().getSimpleName());
  }
}

class CreatorGeneric {
  public static void main(String[] args) {
    Creator c = new Creator();
    c.f();
  }
} /* Output:
X
*///:~












